pkg/controller/cronjob/cronjob_controllerv2.go

```diff
@@ -0,0 +1,647 @@
+		// is already pending deletion. Prevent the job from being a creation observation.
+		jm.deleteJob(job)
+		return
+	}
+
+	// If it has a ControllerRef, that's all that matters.
+	if controllerRef := metav1.GetControllerOf(job); controllerRef != nil {
+		cronJob := jm.resolveControllerRef(job.Namespace, controllerRef)
+		if cronJob == nil {
+			return
+		}
+		jm.enqueueController(cronJob)
+		return
+	}
+}
+
+// updateJob figures out what CronJob(s) manage a Job when the Job
+// is updated and wake them up. If the anything of the Job have changed, we need to
+// awaken both the old and new CronJob. old and cur must be *batchv1.Job
+// types.
+func (jm *ControllerV2) updateJob(old, cur interface{}) {
+	curJob := cur.(*batchv1.Job)
+	oldJob := old.(*batchv1.Job)
+	if curJob.ResourceVersion == oldJob.ResourceVersion {
+		// Periodic resync will send update events for all known jobs.
+		// Two different versions of the same jobs will always have different RVs.
+		return
+	}
+
+	curControllerRef := metav1.GetControllerOf(curJob)
+	oldControllerRef := metav1.GetControllerOf(oldJob)
+	controllerRefChanged := !reflect.DeepEqual(curControllerRef, oldControllerRef)
+	if controllerRefChanged && oldControllerRef != nil {
+		// The ControllerRef was changed. Sync the old controller, if any.
+		if cronJob := jm.resolveControllerRef(oldJob.Namespace, oldControllerRef); cronJob != nil {
+			jm.enqueueController(cronJob)
+		}
+	}
+
+	// If it has a ControllerRef, that's all that matters.
+	if curControllerRef != nil {
+		cronJob := jm.resolveControllerRef(curJob.Namespace, curControllerRef)
+		if cronJob == nil {
+			return
+		}
+		jm.enqueueController(cronJob)
+		return
+	}
+}
+
+func (jm *ControllerV2) deleteJob(obj interface{}) {
+	job, ok := obj.(*batchv1.Job)
+
+	// When a delete is dropped, the relist will notice a job in the store not
+	// in the list, leading to the insertion of a tombstone object which contains
+	// the deleted key/value. Note that this value might be stale.
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("couldn't get object from tombstone %#v", obj))
+			return
+		}
+		job, ok = tombstone.Obj.(*batchv1.Job)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("tombstone contained object that is not a ReplicaSet %#v", obj))
+			return
+		}
+	}
+
+	controllerRef := metav1.GetControllerOf(job)
+	if controllerRef == nil {
+		// No controller should care about orphans being deleted.
+		return
+	}
+	cronJob := jm.resolveControllerRef(job.Namespace, controllerRef)
+	if cronJob == nil {
+		return
+	}
+	jm.enqueueController(cronJob)
+}
+
+func (jm *ControllerV2) enqueueController(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %+v: %v", obj, err))
+		return
+	}
+
+	jm.queue.Add(key)
+}
+
+func (jm *ControllerV2) enqueueControllerWithTime(obj interface{}, t time.Duration) {
```

Comment:
nit: for symmetry with workqueue `enqueueControllerAfter`

Reply:
+1, updated as per suggestion